common/src/main/java/io/pravega/common/tracing/RequestTracker.java

```diff
@@ -0,0 +1,182 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.tracing;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * This class is a helper for multiple classes to get the client-generated id for a request (requestId) based on the
+ * available information of the request itself (requestDescriptor) in multiple methods. The objective is to cache such
+ * [requestDescriptor, requestId] pairs to we allow classes within a component (Controller, Segment Store) to also log
+ * the client-generated ids, making it easier to trace the lifecycle of a request across multiple logs.
+ */
+@Slf4j
+public final class RequestTracker {
+
+    private static final RequestTracker INSTANCE = new RequestTracker();
+    private static final String INTER_FIELD_DELIMITER = "-";
+
+    private final Cache<String, List<Long>> ongoingRequests;
+
+    private RequestTracker() {
+        // Clean request tags after a certain amount of time.
+        ongoingRequests = CacheBuilder.newBuilder()
+                                      .maximumSize(10000)
```

Comment:
How did you select the size? Is 10k really enough for a system that is supposed to be processing tens or hundreds of thousands of operations per second?

Reply:
This value means that a single Controller instance may handle `10K` _stream operations_ (e.g., `createStream`, `deleteStream`) in parallel (not writes/appends). I thought this to be a reasonable (but perhaps conservative) value, considering that requests are normally served in the order of milliseconds. Anyway, I increased that value to match the one already used by `AutoScaleProcessor`.